File: oxidize-golang/core/simd/simd.go
// Package simd mirrors oxidize_core::compute::simd. It enumerates the SIMD
// backends detected at runtime and provides a preferred-backend selection.
package simd

import (
	"runtime"
	"sync"
)

// Backend identifies a SIMD instruction set.
type Backend uint8

const (
	BackendScalar Backend = iota
	BackendSse2
	BackendAvx
	BackendAvx2
	BackendAvx512f
	BackendNeon
)

// String returns the canonical name of the SIMD backend.
func (b Backend) String() string {
	switch b {
	case BackendScalar:
		return "scalar"
	case BackendSse2:
		return "sse2"
	case BackendAvx:
		return "avx"
	case BackendAvx2:
		return "avx2"
	case BackendAvx512f:
		return "avx512f"
	case BackendNeon:
		return "neon"
	default:
		return "unknown"
	}
}

// LaneWidthF32 returns the number of f32 lanes processed per instruction.
func (b Backend) LaneWidthF32() int {
	switch b {
	case BackendScalar:
		return 1
	case BackendSse2, BackendNeon:
		return 4
	case BackendAvx, BackendAvx2:
		return 8
	case BackendAvx512f:
		return 16
	default:
		return 1
	}
}

var (
	detectOnce sync.Once
	detected   []Backend
	preferred  Backend
)

// Available returns the list of backends detected at runtime. The result
// always includes BackendScalar.
func Available() []Backend {
	detectOnce.Do(detect)
	return append([]Backend(nil), detected...)
}

// Preferred returns the highest-performing backend available on this host.
func Preferred() Backend {
	detectOnce.Do(detect)
	return preferred
}

func detect() {
	detected = []Backend{BackendScalar}
	arch := runtime.GOARCH
	switch arch {
	case "amd64":
		// In a real build we'd use runtime.GOARCH + CPUID. We conservatively
		// mark SSE2/AVX/AVX2/AVX512 as available; production deployments
		// should refine this with cpuid intrinsics.
		detected = append(detected, BackendSse2, BackendAvx, BackendAvx2, BackendAvx512f)
		preferred = BackendAvx512f
	case "arm64":
		detected = append(detected, BackendNeon)
		preferred = BackendNeon
	default:
		preferred = BackendScalar
	}
}

File: oxidize-python/oxidize_python/core/simd/simd.py
"""SIMD backend detection mirroring oxidize-golang/core/simd/simd.go."""

from __future__ import annotations

import platform
import threading
from enum import IntEnum


class Backend(IntEnum):
    SCALAR = 0
    SSE2 = 1
    AVX = 2
    AVX2 = 3
    AVX512F = 4
    NEON = 5

    def __str__(self) -> str:
        names = {
            Backend.SCALAR: "scalar",
            Backend.SSE2: "sse2",
            Backend.AVX: "avx",
            Backend.AVX2: "avx2",
            Backend.AVX512F: "avx512f",
            Backend.NEON: "neon",
        }
        return names.get(self, "unknown")

    def lane_width_f32(self) -> int:
        match self:
            case Backend.SCALAR:
                return 1
            case Backend.SSE2 | Backend.NEON:
                return 4
            case Backend.AVX | Backend.AVX2:
                return 8
            case Backend.AVX512F:
                return 16
            case _:
                return 1


_detect_once = threading.Lock()
_detected: list[Backend] | None = None
_preferred: Backend | None = None


def _detect() -> None:
    global _detected, _preferred
    _detected = [Backend.SCALAR]
    arch = platform.machine().lower()
    if arch in ("x86_64", "amd64"):
        _detected.extend([Backend.SSE2, Backend.AVX, Backend.AVX2, Backend.AVX512F])
        _preferred = Backend.AVX512F
    elif arch in ("aarch64", "arm64"):
        _detected.append(Backend.NEON)
        _preferred = Backend.NEON
    else:
        _preferred = Backend.SCALAR


def available() -> list[Backend]:
    global _detected
    if _detected is None:
        with _detect_once:
            if _detected is None:
                _detect()
    return list(_detected or [Backend.SCALAR])


def preferred() -> Backend:
    global _preferred
    if _preferred is None:
        with _detect_once:
            if _preferred is None:
                _detect()
    return _preferred or Backend.SCALAR
